### PR TITLE
ViewRefactor: Fix issues with view refactor in Sacado

### PR DESCRIPTION
**Pull request by crtrott**
Somehow the nested dimensions wouldn't be correclty represented in strides, and DynRankView wasn't accounting for all the specialized helper things in Sacado.

There is a few nasty things here: DynRankView can either take the hidden dimension as an additional argument to its constructor, or pass it through the ViewAllocation properties, via a specialization of these properties in Sacado. Also the DynRankDimTraits are specialized etc. 

It's all fairly messy, but maybe we can get rid of all of it after the full View refactor by rewriting Sacado via using Accessors.

### CODE REVIEW

File: containers/src/Kokkos_DynRankView.hpp [hunk 1]
```diff
@@ -921,9 +929,14 @@ KOKKOS_INLINE_FUNCTION auto subdynrankview(
...
-                     typename sub_t::device_type,
-                     typename sub_t::memory_traits>(sub, new_rank);
+
+  using return_type =
+      DynRankView<typename sub_t::value_type, Kokkos::LayoutStride,
+                  typename sub_t::device_type, typename sub_t::memory_traits>;
+  return static_cast<return_type>(
+      DynRankView<typename sub_t::value_type, typename sub_t::array_layout,
+                  typename sub_t::device_type, typename sub_t::memory_traits>(
+          sub, new_rank));
```
Comment (masterleinad) [inline containers/src/Kokkos_DynRankView.hpp:939]: Why did you need to change `return_type`?

**Review by romintomasetti — COMMENTED** [review]
Do you think it's possible to come up with a test that somehow mimics `Sacado`'s use case ?

File: containers/src/Kokkos_DynRankView.hpp (hunk 1 above)
Comment (crtrott) [reply containers/src/Kokkos_DynRankView.hpp:939]: Because DynRankView used to only ever return LayoutStride in subviews, and some code downstream was relying on that (doing ternary expression with different dimensionality subview, which resulted in different types).

**Comment by crtrott** [comment]
Yes but the test would be roughly the size of Sacado. Essentially Sacado partially specializes almost every implementation detail of Kokkos::View. And as soon as the View refactor is done all these details are gonna be gone and we switch to much cleaner accessors to achieve the same goal. I don't believe it's useful for us to work out those tests. What we could to is an add a CI pass doing Trilinos with just Sacado enabled. It's not terribly long to compile. I will put our folks on that as soon as we have figured out the new testing mechanism on Sandia's machines.

File: containers/src/Kokkos_DynRankView.hpp (hunk 1 above)
Comment (masterleinad) [reply containers/src/Kokkos_DynRankView.hpp:939]: If that's something to change later again (after refactoring Sacado) we should add a comment to that extent.

Comment (crtrott) [reply containers/src/Kokkos_DynRankView.hpp:939]: Its not actually Sacado its in Intrepid2

Comment (crtrott) [reply containers/src/Kokkos_DynRankView.hpp:939]: We can make that part of the DynRank Review later.

**Review by masterleinad — APPROVED** [review]

**Review by janciesko — APPROVED** [review]

**Comment by ndellingwood** [comment]
Tested this PR to check if results of failing tests in https://trilinos-cdash.sandia.gov/viewTest.php?onlyfailed&buildid=1861613 were resolved, only 3 failures (big improvement, thanks @crtrott !):

```
	3122 - PanzerAdaptersSTK_tQuad8ToQuad4Factory_MPI_2 (Failed)
	3123 - PanzerAdaptersSTK_tQuadraticToLinearMeshFactory_MPI_2 (Failed)
	3159 - PanzerAdaptersSTK_tPointBasisValuesEvaluator_MPI_1 (Failed)
```

I'm rebuilding Panzer to compare those tests against a Trilinos build without kokkos@develop to further triage

**Comment by ndellingwood** [comment]
The first two tests failed with Trilinos@develop (not correlated with kokkos@develop changes), however `PanzerAdaptersSTK_tPointBasisValuesEvaluator_MPI_1` passes with kokkos@develop and needs further investigation

**Review by ndellingwood — APPROVED** [review]
I confirmed most of the failing tests from e.g. https://trilinos-cdash.sandia.gov/viewTest.php?onlyfailed&buildid=1861613 pass with this PR
One failing test `PanzerAdaptersSTK_tPointBasisValuesEvaluator_MPI_1` needs more investigation

**Comment by ndellingwood** [comment]
@crtrott I think this is ready for merge?

**Comment by crtrott** [comment]
Failed with a GPU issue one test. Will merge.
